src/databricks/labs/ucx/source_code/graph.py

```diff
@@ -589,6 +588,37 @@ def finalize(self) -> InheritedContext:
         return InheritedContext(tree, self.found)
 
 
+class Lineage(abc.ABC):
```

Comment:
why not a dataclass for typesafety? this cleanly fits all intended use-cases and is more type-safe. Alternatively, we can create `TaskAtom`, `JobAtom`, `WorkspaceAtom`, `PipelineAtom`, `DirectFsAtom`, `TableAtom`, etc etc - for extra compile-time type safety. `object_type` being returned from `as_dict` is a _determinant_ of deserialization type, as we'd need to do some convex hulling or dashboard linking. and we have convention to use `as_dict` instead of `to_json`, as `json` is too ambiguous here.

```python
@dataclass
class Atom:
   object_type: str
   object_id: str
   other: dict[str, str] = field(defaultfactory=dict)

   @staticmethod
   def of(object_type: str, object_id: str, **other: dict[str,str]) -> 'Atom':
       return Atom(object_type, object_type, other)

   def as_dict(self) -> dict[str,str]:
      return {"object_type": self.object_type, "object_id": self.object_id, **self.other}
```

Reply:
That works well as an output, but not so well as an intermediate construct. How do you fit a list of atoms in the above ?

Reply:
done